File: backend/src/main/java/com/studentmanagement/service/StudentService.java
package com.studentmanagement.service;

import com.studentmanagement.model.Student;
import com.studentmanagement.exception.ResourceNotFoundException;
import com.studentmanagement.exception.DuplicateResourceException;

import java.util.List;
import java.util.Map;

/**
 * Service interface for Student operations.
 * This interface demonstrates the Service layer pattern and defines the contract
 * for business logic operations.
 * 
 * Benefits of using an interface:
 * - Abstraction: Hides implementation details
 * - Testability: Easy to create mock implementations for testing
 * - Flexibility: Can have multiple implementations
 */
public interface StudentService {

    /**
     * Retrieve all students from the database.
     * 
     * @return List of all students
     */
    List<Student> getAllStudents();

    /**
     * Retrieve a single student by their ID.
     * 
     * @param id The ID of the student to retrieve
     * @return The student with the given ID
     * @throws ResourceNotFoundException if student is not found
     */
    Student getStudentById(Long id);

    /**
     * Create a new student.
     * 
     * @param student The student object to create
     * @return The created student with generated ID
     * @throws DuplicateResourceException if email already exists
     */
    Student createStudent(Student student);

    /**
     * Update an existing student (full update).
     * 
     * @param id The ID of the student to update
     * @param studentDetails The new student details
     * @return The updated student
     * @throws ResourceNotFoundException if student is not found
     * @throws DuplicateResourceException if email conflicts with another student
     */
    Student updateStudent(Long id, Student studentDetails);

    /**
     * Partially update a student (only specified fields).
     * 
     * @param id The ID of the student to update
     * @param updates Map containing field names and their new values
     * @return The updated student
     * @throws ResourceNotFoundException if student is not found
     * @throws DuplicateResourceException if email conflicts with another student
     */
    Student partialUpdateStudent(Long id, Map<String, Object> updates);

    /**
     * Delete a student by their ID.
     * 
     * @param id The ID of the student to delete
     * @throws ResourceNotFoundException if student is not found
     */
    void deleteStudent(Long id);
}

File: frontend/src/services/studentService.ts
import axios from 'axios';
import type { Student, StudentFormData } from '../types/Student';

/**
 * Base URL for the backend API.
 * Change this if your backend runs on a different port or host.
 */
const API_BASE_URL = 'http://localhost:8080/api/students';

/**
 * Axios instance with default configuration.
 * Includes base URL and default headers.
 */
const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

/**
 * Student Service - Handles all HTTP requests to the backend API.
 * Implements all CRUD operations: Create, Read, Update, Delete.
 */
const studentService = {
  /**
   * GET /api/students
   * Retrieve all students from the database.
   * 
   * @returns Promise with array of all students
   */
  getAllStudents: async (): Promise<Student[]> => {
    const response = await apiClient.get<Student[]>('/');
    return response.data;
  },

  /**
   * GET /api/students/{id}
   * Retrieve a single student by their ID.
   * 
   * @param id - The ID of the student to retrieve
   * @returns Promise with the student data
   */
  getStudentById: async (id: number): Promise<Student> => {
    const response = await apiClient.get<Student>(`/${id}`);
    return response.data;
  },

  /**
   * POST /api/students
   * Create a new student.
   * 
   * @param student - The student data to create
   * @returns Promise with the created student (includes generated ID)
   */
  createStudent: async (student: StudentFormData): Promise<Student> => {
    const response = await apiClient.post<Student>('/', student);
    return response.data;
  },

  /**
   * PUT /api/students/{id}
   * Update an existing student (full update - all fields required).
   * 
   * @param id - The ID of the student to update
   * @param student - The complete student data
   * @returns Promise with the updated student
   */
  updateStudent: async (id: number, student: StudentFormData): Promise<Student> => {
    const response = await apiClient.put<Student>(`/${id}`, student);
    return response.data;
  },

  /**
   * PATCH /api/students/{id}
   * Partially update a student (only specified fields are updated).
   * 
   * @param id - The ID of the student to update
   * @param updates - Object containing only the fields to update
   * @returns Promise with the updated student
   */
  partialUpdateStudent: async (id: number, updates: StudentFormData): Promise<Student> => {
    const response = await apiClient.patch<Student>(`/${id}`, updates);
    return response.data;
  },

  /**
   * DELETE /api/students/{id}
   * Delete a student by their ID.
   * 
   * @param id - The ID of the student to delete
   * @returns Promise that resolves when deletion is complete
   */
  deleteStudent: async (id: number): Promise<void> => {
    await apiClient.delete(`/${id}`);
  },
};

export default studentService;
